File: src/domain/services/SupplySearchService.ts
import { normalizeRouteCode } from '@/domain/value-objects/RouteCode'

export const SUPPLY_CODE_HEAD = '12'
export const SUPPLY_CODE_LENGTH = 11
export const SUPPLY_SEARCH_MIN_DIGITS = 2

export function formatRouteCode(code: string): string {
  const digits = normalizeRouteCode(code)
  if (!digits) return code
  return digits.replace(/\B(?=(\d{3})+(?!\d))/g, ' ')
}

export function expandSupplySearch(query: string): {
  digits: string
  prefixes: string[]
  exactCodes: string[]
} {
  const digits = normalizeRouteCode(query)
  const prefixes = new Set<string>()
  const exactCodes = new Set<string>()

  if (digits.length < SUPPLY_SEARCH_MIN_DIGITS || digits.length > 12) {
    return { digits, prefixes: [], exactCodes: [] }
  }

  prefixes.add(digits)

  const withHead = digits.startsWith(SUPPLY_CODE_HEAD)
    ? digits
    : `${SUPPLY_CODE_HEAD}${digits}`
  if (withHead.length >= SUPPLY_SEARCH_MIN_DIGITS && withHead.length <= 12) {
    prefixes.add(withHead)
  }

  if (digits.length >= 7 && digits.length <= 12) {
    exactCodes.add(digits)
  }
  if (withHead.length >= 7 && withHead.length <= 12) {
    exactCodes.add(withHead)
  }

  if (digits.length >= 4 && digits.length <= 9) {
    const rest = (digits.startsWith(SUPPLY_CODE_HEAD)
      ? digits.slice(SUPPLY_CODE_HEAD.length)
      : digits
    ).padStart(SUPPLY_CODE_LENGTH - SUPPLY_CODE_HEAD.length, '0')
    exactCodes.add(`${SUPPLY_CODE_HEAD}${rest}`.slice(0, SUPPLY_CODE_LENGTH))
  }

  return {
    digits,
    prefixes: [...prefixes],
    exactCodes: [...exactCodes],
  }
}

export function scoreSupplyCode(code: string, digits: string): number {
  if (!digits) return 99
  const value = normalizeRouteCode(code)
  if (!value) return 99
  if (value === digits || value === `${SUPPLY_CODE_HEAD}${digits}`) return 0
  if (value.startsWith(digits) || value.startsWith(`${SUPPLY_CODE_HEAD}${digits}`)) {
    return 1
  }
  if (value.endsWith(digits)) return 2
  if (value.includes(digits)) return 3
  return 10
}

export function supplyCodeMatchesQuery(code: string, query: string): boolean {
  const digits = normalizeRouteCode(query)
  if (!digits) return true
  return scoreSupplyCode(code, digits) < 10
}

export function rankSupplyCodes<T>(
  items: T[],
  digits: string,
  getCode: (item: T) => string,
): T[] {
  return [...items].sort((left, right) => {
    const leftScore = scoreSupplyCode(getCode(left), digits)
    const rightScore = scoreSupplyCode(getCode(right), digits)
    if (leftScore !== rightScore) return leftScore - rightScore
    return getCode(left).localeCompare(getCode(right))
  })
}

export function findCodeHighlight(
  code: string,
  query: string,
): { start: number; end: number } | null {
  const digits = normalizeRouteCode(query)
  if (!digits) return null
  const value = normalizeRouteCode(code)
  const candidates = [digits, `${SUPPLY_CODE_HEAD}${digits}`]
  for (const token of candidates) {
    const index = value.indexOf(token)
    if (index >= 0) return { start: index, end: index + token.length }
  }
  return null
}

File: consorcio/lib/domain/services/supply_search_service.dart
const supplyCodeHead = '12';
const supplyCodeLength = 11;
const supplySearchMinDigits = 2;

class SupplySearchExpansion {
  const SupplySearchExpansion({
    required this.digits,
    required this.prefixes,
    required this.exactCodes,
  });

  final String digits;
  final List<String> prefixes;
  final List<String> exactCodes;
}

String formatRouteCode(String code) {
  final digits = code.replaceAll(RegExp(r'\D'), '');
  if (digits.isEmpty) return code;
  final buffer = StringBuffer();
  final start = digits.length % 3;
  if (start > 0) {
    buffer.write(digits.substring(0, start));
  }
  for (var index = start; index < digits.length; index += 3) {
    if (buffer.isNotEmpty) buffer.write(' ');
    buffer.write(digits.substring(index, index + 3));
  }
  return buffer.toString();
}

SupplySearchExpansion expandSupplySearch(String query) {
  final digits = query.replaceAll(RegExp(r'\D'), '');
  if (digits.length < supplySearchMinDigits || digits.length > 12) {
    return SupplySearchExpansion(
      digits: digits,
      prefixes: const [],
      exactCodes: const [],
    );
  }

  final prefixes = <String>{digits};
  final exactCodes = <String>{};
  final withHead = digits.startsWith(supplyCodeHead)
      ? digits
      : '$supplyCodeHead$digits';
  if (withHead.length >= supplySearchMinDigits && withHead.length <= 12) {
    prefixes.add(withHead);
  }
  if (digits.length >= 7 && digits.length <= 12) {
    exactCodes.add(digits);
  }
  if (withHead.length >= 7 && withHead.length <= 12) {
    exactCodes.add(withHead);
  }
  if (digits.length >= 4 && digits.length <= 9) {
    final rest = (digits.startsWith(supplyCodeHead)
            ? digits.substring(supplyCodeHead.length)
            : digits)
        .padLeft(supplyCodeLength - supplyCodeHead.length, '0');
    final padded = '$supplyCodeHead$rest';
    exactCodes.add(
      padded.length > supplyCodeLength
          ? padded.substring(0, supplyCodeLength)
          : padded,
    );
  }

  return SupplySearchExpansion(
    digits: digits,
    prefixes: prefixes.toList(),
    exactCodes: exactCodes.toList(),
  );
}

int scoreSupplyCode(String code, String digits) {
  if (digits.isEmpty) return 99;
  final value = code.replaceAll(RegExp(r'\D'), '');
  if (value.isEmpty) return 99;
  if (value == digits || value == '$supplyCodeHead$digits') return 0;
  if (value.startsWith(digits) || value.startsWith('$supplyCodeHead$digits')) {
    return 1;
  }
  if (value.endsWith(digits)) return 2;
  if (value.contains(digits)) return 3;
  return 10;
}

bool supplyCodeMatchesQuery(String code, String query) {
  final digits = query.replaceAll(RegExp(r'\D'), '');
  if (digits.isEmpty) return true;
  return scoreSupplyCode(code, digits) < 10;
}
